### PR TITLE
ENH: Adds reverse function for GEOS >= 3.7

### LINKED ISSUE #243: ENH: Add GEOSReverse
Listed in #126, available for GEOS >= 3.7.0

Requested in [geopandas #1682](https://github.com/geopandas/geopandas/issues/1682)

### PR DESCRIPTION
**Pull request by brendan-ward**
Resolves #243

### CODE REVIEW

File: pygeos/constructive.py [hunk 1]
```diff
@@ -438,6 +441,36 @@ def point_on_surface(geometry, **kwargs):
     """
     return lib.point_on_surface(geometry, **kwargs)
 
+
+@requires_geos("3.7.0")
+@multithreading_enabled
+def reverse(geometry, **kwargs):
+    """Returns a copy of the Geometry with the order of coordinates reversed.  If the
+    Geometry is a polygon with interior rings, the interior rings are also reversed.
+
```
Comment (jorisvandenbossche) [inline pygeos/constructive.py:450]:
```suggestion
    """Returns a copy of the Geometry with the order of coordinates reversed.

    If the Geometry is a polygon with interior rings, the interior rings are also
    reversed.
```

(have the first one-liner of the docstring be the main summary. The inner rings note is already more a detail I think)

File: pygeos/constructive.py [hunk 2]
```diff
@@ -438,6 +441,36 @@ def point_on_surface(geometry, **kwargs):
...
+    --------
+    is_ccw : Checks if a Geometry is clockwise.
+
+    Examples
+    --------
+    >>> reverse(Geometry("LINESTRING (0 0, 1 2)"))
+    <pygeos.Geometry LINESTRING (1 2, 0 0)>
+    >>> reverse(Geometry("POLYGON ((0 0, 1 0, 1 1, 0 1, 0 0))"))
+    <pygeos.Geometry POLYGON ((0 0, 0 1, 1 1, 1 0, 0 0))>
+    >>> reverse(None) == None
```
Comment (jorisvandenbossche) [inline pygeos/constructive.py:467]:
```suggestion
    >>> reverse(None) is None
```

File: pygeos/test/test_constructive.py [hunk 3]
```diff
@@ -223,3 +236,73 @@ def test_offset_curve_non_scalar_kwargs():
...
+        (empty_polygon, empty_polygon),
+    ],
+)
+def test_reverse(geom, expected):
+    assert pygeos.equals(pygeos.reverse(geom), expected)
+
+
+@pytest.mark.skipif(pygeos.geos_version < (3, 7, 0), reason="GEOS < 3.7")
+def test_reverse_none():
+    assert pygeos.reverse(None) == None
```
Comment (jorisvandenbossche) [inline pygeos/test/test_constructive.py:294]:
```suggestion
    assert pygeos.reverse(None) is None
```

File: pygeos/test/test_constructive.py [hunk 4]
```diff
@@ -223,3 +236,73 @@ def test_offset_curve_non_scalar_kwargs():
...
+    assert pygeos.reverse([None]).tolist() == [None]
+
+    geometry = pygeos.Geometry("POLYGON ((0 0, 1 0, 1 1, 0 1, 0 0))")
+    expected = pygeos.Geometry("POLYGON ((0 0,  0 1, 1 1, 1 0, 0 0))")
+    result = pygeos.reverse([None, geometry])
+    assert result[0] == None
```
Comment (jorisvandenbossche) [inline pygeos/test/test_constructive.py:300]:
```suggestion
    assert result[0] is None
```

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks! A few small minor nitpicks

**Review by jorisvandenbossche — APPROVED** [review]

File: pygeos/constructive.py [hunk 5]
```diff
@@ -438,6 +441,38 @@ def point_on_surface(geometry, **kwargs):
...
+
+@requires_geos("3.7.0")
+@multithreading_enabled
+def reverse(geometry, **kwargs):
+    """Returns a copy of the Geometry with the order of coordinates reversed.
+
+    If the Geometry is a polygon with interior rings, the interior rings are also
+    reversed.
+
+    Points are unchanged.  None is returned when Geometry is None.
```
Comment (caspervdw) [inline pygeos/constructive.py:453]:
```suggestion
    Points are unchanged. None is returned where Geometry is None.
```

File: pygeos/constructive.py [hunk 6]
```diff
@@ -438,6 +441,38 @@ def point_on_surface(geometry, **kwargs):
     """
     return lib.point_on_surface(geometry, **kwargs)
 
+
+@requires_geos("3.7.0")
+@multithreading_enabled
+def reverse(geometry, **kwargs):
+    """Returns a copy of the Geometry with the order of coordinates reversed.
```
Comment (caspervdw) [inline pygeos/constructive.py:448]:
```suggestion
    """Returns a copy of a Geometry with the order of coordinates reversed.
```

**Review by caspervdw — APPROVED** [review]
Two minor docstring comments. I always try to address not address the function input as "the" geometry as it implies it is singular.

+ Don't forget to append the changelog before merging.

**Pull request merged**